components/canvas/canvas_paint_thread.rs

```diff
@@ -2,1137 +2,280 @@
+            Canvas2dMsg::SetLineJoin(join) => {
+                self.canvas(&canvas_id).set_line_join(join)
+            },
+            Canvas2dMsg::SetMiterLimit(limit) => {
+                self.canvas(&canvas_id).set_miter_limit(limit)
+            },
+            Canvas2dMsg::SetTransform(ref matrix) => {
+                self.canvas(&canvas_id).set_transform(matrix)
+            },
+            Canvas2dMsg::SetGlobalAlpha(alpha) => {
+                self.canvas(&canvas_id).set_global_alpha(alpha)
+            },
+            Canvas2dMsg::SetGlobalComposition(op) => {
+                self.canvas(&canvas_id).set_global_composition(op)
+            },
+            Canvas2dMsg::GetImageData(dest_rect, canvas_size, chan) => {
+                self.canvas(&canvas_id).image_data(dest_rect, canvas_size, chan)
+            },
+            Canvas2dMsg::PutImageData(
+                imagedata,
+                offset,
+                image_data_size,
+                dirty_rect,
+            ) => {
+                self.canvas(&canvas_id).put_image_data(
+                    imagedata.into(),
+                    offset,
+                    image_data_size,
+                    dirty_rect,
+                )
+            },
+            Canvas2dMsg::SetShadowOffsetX(value) => {
+                self.canvas(&canvas_id).set_shadow_offset_x(value)
+            },
+            Canvas2dMsg::SetShadowOffsetY(value) => {
+                self.canvas(&canvas_id).set_shadow_offset_y(value)
+            },
+            Canvas2dMsg::SetShadowBlur(value) => {
+                self.canvas(&canvas_id).set_shadow_blur(value)
+            },
+            Canvas2dMsg::SetShadowColor(ref color) => {
+                self.canvas(&canvas_id).set_shadow_color(color.to_azure_style())
             },
-            FillOrStrokeStyle::Surface(ref surface_style) => {
-                drawtarget.create_source_surface_from_data(&surface_style.surface_data,
-                                                           surface_style.surface_size,
-                                                           surface_style.surface_size.width * 4,
-                                                           SurfaceFormat::B8G8R8A8)
-                          .map(|source_surface| {
-                    Pattern::Surface(SurfacePattern::new(
-                        source_surface.azure_source_surface,
-                        surface_style.repeat_x,
-                        surface_style.repeat_y,
-                        &Transform2D::identity()))
-                    })
-            }
         }
     }
-}
-
-impl ToAzureStyle for RGBA {
-    type Target = Color;
 
-    fn to_azure_style(self) -> Color {
-        Color::rgba(self.red_f32() as AzFloat,
-                    self.green_f32() as AzFloat,
-                    self.blue_f32() as AzFloat,
-                    self.alpha_f32() as AzFloat)
+    fn canvas(&mut self, canvas_id: &CanvasId) -> &mut CanvasData<'a> {
```

Comment:
If you make `CanvasId` copy, you don't need to add references to avoid `clone()`, which would read nicer IMO.

Reply:
Also done.